### ISSUE TITLE
Please support mounting NBD disks

### ISSUE THREAD

**Issue by joshtriplett** [microsoft/WSL#5968]
The Linux kernel supports Network Block Device (NBD) disks. I'd like to be able to mount such disks, containing any filesystem Linux supports, and access them from both Linux and Windows.

This is a capability easily available on a native Linux system, and it would be helpful to have the same capability on WSL2.

**Comment by guilt** [comment]
Please enable it. I use BUSE for testing block corruption, and this would be very helpful.

CONFIG_BLK_DEV_NBD=y is still disabled, so that needs to go in for nbd-client and nbd-server to work. I've manually put together a kernel so I can do this, but would really appreciate this to be out of the box.

**Comment by pojntfx** [comment]
I second this. Docker now uses WSL2 by default, and tools such as [alpimager](https://github.com/pojntfx/alpimager) start to fail on Docker for Windows because nbd support is lacking.

**Comment by samliddicott** [comment]
This was going to be my way to get around `wsl --mount`  limitation:

* At this time, only entire disks can be attached to WSL 2, meaning that it's not possible to attach only a partition. Concretely, this means that it's not possible to use wsl --mount to read a partition on the boot device, because that device can't be detached from Windows.

**Comment by samliddicott** [comment]
Work-around in the meantime; build your own kernel: https://blog.dan.drown.org/replacing-the-wsl-kernel/

**Comment by samliddicott** [comment]
It ought the be practically the same as the Ubuntu kernel right out of the box.

I'm now stuck on accessing the windows NBD server from Ubuntu.

**Comment by NathanNZ** [comment]
Any progress on this? Would be great to have CONFIG_BLK_DEV_NBD set to "y" by default!

**Comment by Biswa96** [comment]
You can compile and use your own Linux kernel in WSL2 with that config enabled.

**Comment by joshtriplett** [comment]
Building a custom kernel has been mentioned multiple times in this issue, and people asking for this know that that's an option. It would be helpful to have NBD work out of the box.

**Comment by samliddicott** [comment]
Slightly (very?) off topic, but I found it much simpler to switch to native
linux. With outlook and office being online anyway, it wasn't too hard a
switch. It's much faster too.  It's not the answer you want, and it's not
the one I wanted, but it's a lot less faff all round.

**Comment by ShadowJonathan** [comment]
Why is this not yet supported? Today i had to quickly verify a qcow2 image on a WD Passport disk (which only supports windows or mac), which is only mountable via nbd, and I couldn't do this, because WSL2 doesn't support this.

**Comment by matt-psaltis** [comment]
Just adding the annual +1 :)

**Comment by drew-viles** [comment]
I came here hoping to find, well, some movement of any kind on this since it's been open for 2 years 😞 

I guess it's back to dual-booting or running Linux in a VM until Microsoft add some basic features that most Linux admins would love to see in WSL to prevent said dual booting &/or VMs.

It was nice knowing ya WSL, but you ain't quite what I'm looking for in a partner.

¯\_(ツ)_/¯

**Comment by okibcn** [comment]
Why does changing the default config file for building the wsl2 kernel take that long?

That setting works and it is handy for mounting qcow2 images quickly.

**Comment by guilt** [comment]
> Why does changing the default config file for building the wsl2 kernel take that long?

If only the world 😒 worked this way, we'd all be in the year 3000 🚀

**Comment by okibcn** [comment]
If you are interested in having the latest kernel with nbd enabled and always updated to the latest release, you can get it from https://github.com/okibcn/WSL2-linux-kernel/releases/latest

I have added another configuration for enabling settings to run Anbox. The Zip includes all the 3 kernels and an installer/uninstaller.  I created this for myself so I didn't have to manually build it after each update. I also created a scoop manifest to help me get it updated easily.

It is sad Microsoft didn't have it enabled and I had to do this.

**Comment by guilt** [comment]
> It is sad Microsoft didn't have it enabled and I had to do this.

Right. We built our own kernels when we wrote this. We really just wish this is out of the box.

**Comment by okibcn** [comment]
It works quite well:
![ses](https://user-images.githubusercontent.com/22417711/220540357-908b0542-2f66-4ef8-80c8-da6099c38696.gif)

**Comment by drew-viles** [comment]
You don't appear to be using the standard kernel.
Have you built your own?

Ideally this should be implemented in the standard one so we don't have to go adding basic features into the kernel and then be responsible for manually upgrading every time.

```
~# uname -a
Linux Drew-PC 5.15.79.1-microsoft-standard-WSL2 #1 SMP Wed Nov 23 01:01:46 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

~# ls /dev/nbd*
ls: cannot access '/dev/nbd*': No such file or directory`
``

**Comment by okibcn** [comment]
@drew-viles yes, it is my repo. I automated a workflow to build a new nbd version of the kernel as soon as official sources are released by Microsoft. I have been doing it manually before. 
I can now mount qcow2 images with qemu-nbd in WSL without any issues and without having to manually compile the kernel.

It is just a setting, nothing complicated. Really, MS could enable this as it works with WSL without any side effects.

**Comment by jonylentzmc** [comment]
Anual Bump to this

**Comment by naftalysh** [comment]
@ocibcn
Could you please explain how do we use your procedure as you mentioned "I automated a workflow to build a new nbd version of the kernel as soon as official sources are released by Microsoft"?

**Comment by ChenPi11** [comment]
See https://github.com/ChenPi11/WSL
I'd tried NBD before, but I think it's a unstable solution because NBD may broken when machine suspended, and Hyper-V's vsock also reports unstable during high-payload data transfer.
Maybe iSCSI a better solution?

**Comment by Mart-Bogdan** [comment]
Isn't WSL shutting down during suspend? I am not sure about it but I thought this is the case. Interesting,

**Comment by ChenPi11** [comment]
> Isn't WSL shutting down during suspend? I am not sure about it but I thought this is the case. Interesting,

WSL will not shutting down. But nbd will disconnect during suspend. After resumed, any operate on the nbd mount point will cause I/O error. and the broken nbd device will be unable to use forever. (unless reboot)
